File: app/src/main/java/com/neocaptainnemo/cv/ui/common/CommonViewModel.java
package com.neocaptainnemo.cv.ui.common;

import android.text.Html;
import android.text.Spanned;

import com.neocaptainnemo.cv.model.CommonSection;

public class CommonViewModel {

    private CommonSection commonSection;

    public CommonViewModel(CommonSection commonSection) {
        this.commonSection = commonSection;
    }

    public String getTitle() {
        return commonSection.title;
    }

    public Spanned getDescription() {
        return Html.fromHtml(commonSection.description);
    }
}

File: app/src/main/java/com/neocaptainnemo/cv/ui/common/CommonViewModel.kt
package com.neocaptainnemo.cv.ui.common

import androidx.lifecycle.ViewModel
import com.neocaptainnemo.cv.core.data.DataService
import com.neocaptainnemo.cv.core.model.CommonSection
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.onStart
import javax.inject.Inject

@HiltViewModel
class CommonViewModel @Inject constructor(private val dataService: DataService) : ViewModel() {

    private val _empty = MutableStateFlow(false)

    private val _progress = MutableStateFlow(false)

    val empty: Flow<Boolean> = _empty

    val progress: Flow<Boolean> = _progress

    fun commons(): Flow<List<CommonSection>> = dataService.commons()
        .catch {
            emit(listOf())
        }
        .onStart {
            _progress.value = true
        }
        .onEach {
            _progress.value = false
            _empty.value = it.isEmpty()
        }
}
